### PR TITLE
Object favorites

### PR DESCRIPTION
**Pull request by dustincoleman**
Result Provider changes needed to support object favorites. (The ability to pin properties and fields as favorites of a type in the EE windows.)

### CODE REVIEW

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/MemberExpansion.cs [hunk 1]
```diff
@@ -70,11 +72,18 @@ internal sealed class MemberExpansion : Expansion
             var includeInherited = (flags & ExpansionFlags.IncludeBaseMembers) == ExpansionFlags.IncludeBaseMembers;
             var hideNonPublic = (inspectionContext.EvaluationFlags & DkmEvaluationFlags.HideNonPublicMembers) == DkmEvaluationFlags.HideNonPublicMembers;
             var includeCompilerGenerated = (inspectionContext.EvaluationFlags & DkmEvaluationFlags.ShowValueRaw) == DkmEvaluationFlags.ShowValueRaw;
-            runtimeType.AppendTypeMembers(allMembers, predicate, declaredTypeAndInfo.Type, appDomain, includeInherited, hideNonPublic, isProxyType, includeCompilerGenerated);
+            var favoritesInfo = (supportsFavorites) ? type.GetFavorites() : null;
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/MemberExpansion.cs]: >( [](start = 32, length = 1)

nit: braces not necessary

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/MemberExpansion.cs [hunk 2]
```diff
@@ -70,11 +72,18 @@ internal sealed class MemberExpansion : Expansion
...
+            var favoritesMembersByName = new Dictionary<string, MemberAndDeclarationInfo>();
 
             foreach (var member in allMembers)
             {
-                if (member.IsStatic)
+                if (member.IsFavorite)
+                {
+                    favoritesMembersByName.Add(member.Name, member);
+                }
+                else if (member.IsStatic)
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/MemberExpansion.cs]: >else if [](start = 16, length = 7)

Does this mean we do not duplicate favorites when listing regular members? Correct?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/DebuggerDisplayInfo.cs [hunk 3]
```diff
@@ -1,19 +1,80 @@
...
+        }
+
+        internal void ApplyEvalAttribute(DkmClrDebuggerDisplayAttribute attribute, DkmClrType attributeTarget)
+        {
+            if (attribute.Name != null)
+            {
+                Name = new DebuggerDisplayItemInfo(attribute.Name, attributeTarget);
+            }
+
+            if (attribute.Value != null && m_value == null && m_simpleValue == null) // Favorites display string takes priority for Value
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/DebuggerDisplayInfo.cs]: >// Favorites display string takes priority for Value [](start = 85, length = 52)

please put the comment above the code line

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/EvalResultDataItem.cs [hunk 4]
```diff
@@ -193,7 +199,13 @@ internal EvalResultDataItem ToDataItem()
                 FormatSpecifiers);
         }
 
-        private static DkmEvaluationResultFlags GetFlags(DkmClrValue value, DkmInspectionContext inspectionContext)
+        internal void DisableCanAddFavorite()
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/EvalResultDataItem.cs]: >DisableCanAddFavorite [](start = 22, length = 21)

We usually prefer immutable object. The pattern is to create a new object using WithDisableCanAddFavorite()

**Comment by ivanbasov** [comment]
> using System;

copyright

---
Refers to: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/FavoritesDataItem.cs:1 in 594892a. [](commit_id = 594892a4fa79a3df31d234bfa14fac01174524ee, deletion_comment = False)

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/DebuggerDisplayInfo.cs [hunk 5]
```diff
@@ -1,19 +1,80 @@
...
+        {
+            if (m_simpleValue != null && (inspectionContext.EvaluationFlags & DkmEvaluationFlags.UseSimpleDisplayString) == DkmEvaluationFlags.UseSimpleDisplayString)
+            {
+                return m_simpleValue;
+            }
+
+            return m_value;
+        }
+
+        internal void ApplyFavorities(DkmClrObjectFavoritesInfo favoritesInfo)
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/DebuggerDisplayInfo.cs]: >ApplyFavorities [](start = 22, length = 15)

We usually prefer immutable object. The pattern it WithFavoritesInfo(DkmClrObjectFavoritesInfo favoritesInfo). Please let me know if there is a reason to have it mutable.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/DebuggerDisplayInfo.cs [hunk 6]
```diff
@@ -1,19 +1,80 @@
...
+                m_value = new DebuggerDisplayItemInfo(favoritesInfo.DisplayString, m_targetType);
+
+                if (favoritesInfo.SimpleDisplayString != null)
+                {
+                    m_simpleValue = new DebuggerDisplayItemInfo(favoritesInfo.SimpleDisplayString, m_targetType);
+                }
+            }
+        }
+
+        internal void ApplyEvalAttribute(DkmClrDebuggerDisplayAttribute attribute, DkmClrType attributeTarget)
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/DebuggerDisplayInfo.cs]: >ApplyEvalAttribute [](start = 22, length = 18)

Same as above. Is there a reason to have it mutable?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/TypeHelpers.cs [hunk 7]
```diff
@@ -50,6 +52,31 @@ internal static class TypeHelpers
...
+                        if (browsableStateValue == DkmClrDebuggerBrowsableAttributeState.RootHidden)
+                        {
+                            supportsFavorites = false;
+                            break;
+                        }
+                    }
+                }
+
+                // Get the favorites information if it is supported.
+                Dictionary<string, object> favoritesMemberNames = null;
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/TypeHelpers.cs]: >Dictionary<string, object> [](start = 16, length = 26)

it seems that a hashset is enough.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/TypeHelpers.cs [hunk 8]
```diff
@@ -137,7 +164,14 @@ internal static class TypeHelpers
...
 
-                        includedMembers.Add(new MemberAndDeclarationInfo(member, browsableStateValue, previousDeclaration, inheritanceLevel));
+                        includedMembers.Add(
+                            new MemberAndDeclarationInfo(
+                                member,
+                                browsableStateValue,
+                                previousDeclaration,
+                                inheritanceLevel,
+                                canFavorite: supportsFavorites,
+                                isFavorite: (favoritesMemberNames?.ContainsKey(memberName) == true)));
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/TypeHelpers.cs]: >( [](start = 44, length = 1)

braces are not necessary

**Comment by ivanbasov** [comment]
I see a lot of test were update. However, I could not find any new tests verifying the feature. Should we have them?

**Comment by ivanbasov** [comment]
Done with reviewing iteration 7

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/MemberExpansion.cs (hunk 2 above)
Comment (dustincoleman) [reply src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/MemberExpansion.cs]: Yes, favorites "move" to the top of the list.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/TypeHelpers.cs (hunk 7 above)
Comment (dustincoleman) [reply src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/TypeHelpers.cs]: It is, but unfortunately this still has to build for .net 2.0 which does not have it.

**Comment by dustincoleman** [comment]
> > I see a lot of test were update. However, I could not find any new tests verifying the feature. Should we have them?
> 
> The feature is actually very well covered with all of the existing tests. The only thing that is not covered is the new favorites display string logic. I will add some tests for that.

And filtering to favorites only. I'll add some for that as well.

**Comment by dustincoleman** [comment]
> I see a lot of test were update. However, I could not find any new tests verifying the feature. Should we have them?

[Update] Yes, I will add some.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/TypeHelpers.cs (hunk 7 above)
Comment (ivanbasov) [reply src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/TypeHelpers.cs]: Wow.. interesting! 
And should this binary work under .net 2.0 as well?

If so, could you please then add a comment about this and the usage of Dictionary for the purpose?

---
In reply to: [326297622](https://github.com/dotnet/roslyn/pull/38669#discussion_r326297622) [](ancestors = 326297622)

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/DebuggerDisplayInfo.cs [hunk 9]
```diff
@@ -1,19 +1,120 @@
...
 
+using System;
 using Microsoft.VisualStudio.Debugger.Clr;
+using Microsoft.VisualStudio.Debugger.Evaluation;
 using Microsoft.VisualStudio.Debugger.Evaluation.ClrCompilation;
 
 namespace Microsoft.CodeAnalysis.ExpressionEvaluator
 {
-    internal struct DebuggerDisplayInfo
+    internal class DebuggerDisplayInfo
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/DebuggerDisplayInfo.cs]: >class [](start = 13, length = 5)

Should we change this to class?

File: src/ExpressionEvaluator/Core/Test/ResultProvider/Debugger/Engine/DkmClrObjectFavoritesInfo.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            Favorites = new ReadOnlyCollection<string>(favorites);
+            DisplayString = displayString;
+            SimpleDisplayString = simpleDisplayString;
+        }
+
+        public string DisplayString { get; }
+        public string SimpleDisplayString { get; }
+        public ReadOnlyCollection<string> Favorites { get; }
+}
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Test/ResultProvider/Debugger/Engine/DkmClrObjectFavoritesInfo.cs]: nit alignment

File: src/ExpressionEvaluator/Core/Test/ResultProvider/Debugger/Engine/DkmClrRuntimeInstance.cs [hunk 11]
```diff
@@ -48,6 +50,13 @@ public class DkmClrRuntimeInstance : DkmRuntimeInstance
             this.GetMemberValue = getMemberValue;
         }
 
+        internal DkmClrRuntimeInstance(Assembly[] assemblies, Dictionary<string, DkmClrObjectFavoritesInfo> favoritesByTypeName)
+            : this(assemblies)
+        {
+            _favoritesByTypeName = favoritesByTypeName;
+        }
+
+
```
Comment (ivanbasov) [inline src/ExpressionEvaluator/Core/Test/ResultProvider/Debugger/Engine/DkmClrRuntimeInstance.cs]: nit: extra empty line

**Comment by ivanbasov** [comment]
Great! Thank you! Should we add similar tests for VB?

---
In reply to: [533271481](https://github.com/dotnet/roslyn/pull/38669#issuecomment-533271481) [](ancestors = 533271481)

**Comment by ivanbasov** [comment]
Done with reviewing iteration 9. Major question - tests for VB. Other than that ready to signoff.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/DebuggerDisplayInfo.cs (hunk 9 above)
Comment (dustincoleman) [reply src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/DebuggerDisplayInfo.cs]: I suppose we could keep it a value type, but now that it contains logic and private data it seemed more appropriate to be a reference type. It make the initialization logic simpler in the constructors.

**Comment by ivanbasov** [comment]
@dotnet/roslyn-infrastructure please signoff as well: from infrastructure perspective, this PR contains just a version upgrade of debugger binaries. Thank you!

File: src/ExpressionEvaluator/CSharp/Test/ResultProvider/ObjectFavoritesTests.cs [hunk 12]
```diff
@@ -0,0 +1,198 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Linq;
+using DiffPlex.Model;
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Test/ResultProvider/ObjectFavoritesTests.cs]: Is this used?

**Pull request merged**